allium/lib/search_discovery.py

```
@@ -37,17 +38,73 @@ def _route_for_html(relative_path):
-    output_path = Path(output_dir)
-    urls = []
-    for html_path in output_path.rglob("*.html"):
+class _HeadSignalsParser(HTMLParser):
+    """Extract canonical and robots directives without third-party parsers."""
+
+    def __init__(self):
+        super().__init__(convert_charrefs=True)
+        self.canonicals = []
+        self.noindex = False
+
+    def handle_starttag(self, tag, attrs):
```

Ruff (0.16.0)

[warning] Missing return type annotation for private function `handle_starttag`

Add return type annotation: `None`

(ANN202)